onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -1,69 +1,87 @@
-  // Create execution frame for executing constant nodes.
-  OptimizerExecutionFrame::Info info({&node}, graph.GetAllInitializedTensors());
+Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-  std::vector<int> fetch_mlvalue_idxs;
-  for (const auto* node_out : node.OutputDefs()) {
-    fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-  }
+  for (NodeIndex i : order) {
+    auto* node = graph.GetNode(i);
+    if (!node) {
+      return Status(ONNXRUNTIME, INVALID_ARGUMENT);
+    }
 
-  OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
+    // TODO Further investigate if this is needed here.
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level));
 
-  auto* kernel = info.GetKernel(node.Index());
-  OpKernelContext op_kernel_context(&frame, kernel, ::onnxruntime::logging::LoggingManager::DefaultLogger());
+    // Check if constant folding can be applied on this node.
+    if (!graph_utils::IsSupportedProvider(*node, GetCompatibleExecutionProviders()) ||
+        excluded_op_types_.find(node->OpType()) != excluded_op_types_.end() ||
+        // constant folding is not currently supported in subgraphs.
+        node->ImplicitInputDefs().size() != 0 ||
+        !graph_utils::AllNodeInputsAreConstant(graph, *node)) {
+      continue;
+    }
 
-  kernel->Compute(&op_kernel_context);
+    // Create execution frame for executing constant nodes.
+    OptimizerExecutionFrame::Info info({node}, graph.GetAllInitializedTensors());
 
-  std::vector<MLValue> fetches;
-  frame.GetOutputs(fetches);
+    std::vector<int> fetch_mlvalue_idxs;
+    for (const auto* node_out : node->OutputDefs()) {
+      fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
+    }
 
-  // Go over all output node args and substitute them with the newly computed tensors, which will be
-  // added to the graph as initializers.
-  ORT_ENFORCE(fetches.size() == node.OutputDefs().size());
-  for (size_t fetch_idx = 0; fetch_idx < fetches.size(); ++fetch_idx) {
-    MLValue& mlvalue = fetches[fetch_idx];
+    OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
 
-    // Build the TensorProto that corresponds to the computed MLValue and add it as initializer to the graph.
-    ONNX_NAMESPACE::TensorProto out_tensorproto;
-    const auto* constant_arg_out = node.OutputDefs()[fetch_idx];
-    BuildTensorProtoForInitializer(mlvalue, *constant_arg_out, out_tensorproto);
+    auto* kernel = info.GetKernel(node->Index());
+    OpKernelContext op_kernel_context(&frame, kernel, ::onnxruntime::logging::LoggingManager::DefaultLogger());
 
-    graph.AddInitializedTensor(out_tensorproto);
-  }
+    kernel->Compute(&op_kernel_context);
+
+    std::vector<MLValue> fetches;
+    frame.GetOutputs(fetches);
+
+    // Go over all output node args and substitute them with the newly computed tensors, which will be
```

Comment:
I don't see a check to see if the node output is present in the graph output.... is this case implicitly supported?

Reply:
It would be safer to add the check here. We could get to a point that a graph has no nodes otherwise, which I dont think is supported.

Reply:
I added an extra check. It is not clear that a graph without any nodes is supported.
From some small examples I created, it seems that ONNX might support it, but requires further investigation before allowing such cases.